Firestore/Example/Tests/Integration/API/FIRCursorTests.mm

```diff
@@ -192,4 +192,82 @@ - (void)testCanBeUsedInDescendingQueries {
   XCTAssertEqualObjects(FIRQuerySnapshotGetData(snapshot), (@[ @{ @"v" : @"d", @"sort" : @3.0 } ]));
 }
 
+FIRTimestamp *timestampWithMicros(int64_t seconds, int32_t micros) {
+  // Firestore only supports microsecond resolution, so use a microsecond as a minimum value for
+  // nanoseconds.
+  return [FIRTimestamp timestampWithSeconds:seconds nanoseconds:micros * 1000];
+}
+
+- (void)testTimestampsCanBePassedToQueriesAsLimits {
+  FIRCollectionReference *testCollection = [self collectionRefWithDocuments:@{
+    @"a" : @{@"timestamp" : timestampWithMicros(100, 2)},
+    @"b" : @{@"timestamp" : timestampWithMicros(100, 5)},
+    @"c" : @{@"timestamp" : timestampWithMicros(100, 3)},
+    @"d" : @{@"timestamp" : timestampWithMicros(100, 1)},
+    // Number of microseconds deliberately repeated.
+    @"e" : @{@"timestamp" : timestampWithMicros(100, 5)},
+    @"f" : @{@"timestamp" : timestampWithMicros(100, 4)},
+  }];
+  FIRQuery *query = [testCollection queryOrderedByField:@"timestamp"];
+  FIRQuerySnapshot *querySnapshot =
+      [self readDocumentSetForRef:[[query queryStartingAfterValues:@[ timestampWithMicros(100, 2) ]]
+                                      queryEndingAtValues:@[ timestampWithMicros(100, 5) ]]];
+  XCTAssertEqualObjects(FIRQuerySnapshotGetIDs(querySnapshot), (@[ @"c", @"f", @"b", @"e" ]));
+}
+
+- (void)testTimestampsCanBePassedToQueriesInWhereClause {
+  FIRTimestamp *currentTimestamp = [FIRTimestamp timestamp];
+  int64_t seconds = currentTimestamp.seconds;
+  int32_t micros = currentTimestamp.nanoseconds / 1000;
+  FIRCollectionReference *testCollection = [self collectionRefWithDocuments:@{
+    @"a" : @{
+      @"timestamp" : timestampWithMicros(seconds, micros + 2),
+    },
+    @"b" : @{
+      @"timestamp" : timestampWithMicros(seconds, micros - 1),
+    },
+    @"c" : @{
+      @"timestamp" : timestampWithMicros(seconds, micros + 3),
+    },
+    @"d" : @{
+      @"timestamp" : timestampWithMicros(seconds, micros),
+    },
+    @"e" : @{
+      @"timestamp" : timestampWithMicros(seconds, micros + 1),
+    }
+  }];
+
+  FIRQuerySnapshot *querySnapshot = [self
+      readDocumentSetForRef:[[testCollection queryWhereField:@"timestamp"
+                                      isGreaterThanOrEqualTo:timestampWithMicros(seconds, micros)]
+                                queryWhereField:@"timestamp"
+                                     isLessThan:timestampWithMicros(seconds, micros + 3)]];
+  XCTAssertEqualObjects(FIRQuerySnapshotGetIDs(querySnapshot), (@[ @"d", @"e", @"a" ]));
```

Comment:
Looks like the results are ordered by timestamp, even though I didn't specify an explicit ordering.

Reply:
The use of an inequality triggers that behavior. When using an inequality and order bys, the inequality and first order by must match to make the index scan work.